contracts/Staking.sol

```diff
@@ -0,0 +1,246 @@
+        );
+        require(
+            rewardsAmount_ > 0 && start_ > 0 && apr_ > 0,
+            "Error : one of params equal to 0"
+        );
+        stakingParams.stakingStartDate = start_;
+        stakingParams.stakingFinishDate =
+            stakingParams.stakingStartDate +
+            stakingParams.stakingPeriod;
+        apr = apr_;
+        maxStakingPool = (rewardsAmount_ * ONE_HUNDRED_PERCENT) / apr;
+        token.safeTransferFrom(msg.sender, address(this), rewardsAmount_);
+        emit SetRewards(start_, rewardsAmount_, apr_);
+    }
+
+    /**
+     * @dev function allow user stake tokens
+     *
+     * @param amount_ - how many tokens user want to stake
+     *
+     * @notice function can call by any one , but sum of
+     * all staking tokens can't be higher than 'MAX_STAKING_POOL'
+     */
+    function stake(uint256 amount_) external override {
+        require(
+            block.timestamp >
+                investorList[msg.sender].lastTimeStake +
+                    stakingParams.cooldownPeriod,
+            "Error : for re-staking wait 10 days"
+        );
+        require(amount_ > 0, "Error : you can't stake 0 tokens");
+        require(
+            stakingParams.stakingStartDate > 0,
+            "Error : staking has not started yet"
+        );
+        require(
+            stakingParams.stakingFinishDate > block.timestamp,
+            "Error : staking period has end"
+        );
+        _updateReward(msg.sender);
+        investorList[msg.sender].stakingAmount += amount_;
+        require(
+            stakingTotalAmount + amount_ <= maxStakingPool,
+            "Error : your stake value too high"
+        );
+        stakingTotalAmount += amount_;
+        investorList[msg.sender].lastTimeStake = block.timestamp;
+        token.safeTransferFrom(msg.sender, address(this), amount_);
+        emit Stake(msg.sender, amount_);
+    }
+
+    /**
+     * @dev function allow user to withdraw sum of stake tokens
+     * by caller + collected reward for caller
+     *
+     * @notice if the user call this function before the end of staking period
+     * he will receive only 60% of his reward
+     */
+    function unStake() external override {
+        require(
+            investorList[msg.sender].stakingAmount > 0,
+            "Error: you are not investor"
+        );
+        _updateReward(msg.sender);
+        uint256 amount;
+        uint256 reward = investorList[msg.sender].reward;
+        if (block.timestamp < stakingParams.stakingFinishDate) {
+            reward =
+                reward -
+                ((reward * stakingParams.feePercent) / ONE_HUNDRED_PERCENT);
+        }
+        amount = investorList[msg.sender].stakingAmount + reward;
+        stakingTotalAmount -= investorList[msg.sender].stakingAmount;
+        delete investorList[msg.sender];
+        token.safeTransfer(msg.sender, amount);
+        emit UnStake(msg.sender, amount);
+    }
+
```

Comment:
after staking period, some tokens will lock on contract, should add method for get tokens which not will distribute to users